File: examples/quickstarts/openai_responses_api.py
# ---------------------------------------------
## Full example: Use Retab with OpenAI's Responses API and JSON Schema extraction.
# ---------------------------------------------

import os

from dotenv import load_dotenv
from openai import OpenAI

from retab import Schema, Retab
from retab._utils.json_schema import filter_auxiliary_fields_json

# Load environment variables
load_dotenv()

api_key = os.getenv("OPENAI_API_KEY")
retab_api_key = os.getenv("RETAB_API_KEY")

assert api_key, "Missing OPENAI_API_KEY"
assert retab_api_key, "Missing RETAB_API_KEY"

# Define schema
json_schema = {
    'X-SystemPrompt': 'You are a useful assistant extracting information from documents.',
    'properties': {
        'name': {
            'description': 'The name of the calendar event.',
            'title': 'Name',
            'type': 'string',
        },
        'date': {
            'X-ReasoningPrompt': 'The user can mention it in any format, like **next week** or **tomorrow**. Infer the right date format from the user input.',
            'description': 'The date of the calendar event in ISO 8601 format.',
            'title': 'Date',
            'type': 'string',
        },
    },
    'required': ['name', 'date'],
    'title': 'CalendarEvent',
    'type': 'object',
}

# Configuration
model = "gpt-4o"
modality = "native"
temperature = 0.0

# Retab Setup
reclient = Retab(api_key=retab_api_key)
doc_msg = reclient.documents.create_messages(
    document="../../assets/calendar_event.xlsx",
    modality=modality,
    image_resolution_dpi=96,
    browser_canvas="A4",
)
schema_obj = Schema(json_schema=json_schema)

# OpenAI Responses API call
client = OpenAI(api_key=api_key)
# Example 1: With the client.responses.create() method
response = client.responses.create(
    model=model,
    temperature=temperature,
    input=schema_obj.openai_responses_input + doc_msg.openai_responses_input,
    text={"format": {"type": "json_schema", "name": schema_obj.id, "schema": schema_obj.inference_json_schema, "strict": True}},
)

# Example 2: With the client.responses.parse() method
client = OpenAI(api_key=api_key)
response = client.responses.parse(
    model=model,
    temperature=temperature,
    input=schema_obj.openai_responses_input + doc_msg.openai_responses_input,
    text_format=schema_obj.inference_pydantic_model,
)

# Example 3: Put the system prompt in the instructions parameter
client = OpenAI(api_key=api_key)
response = client.responses.parse(
    model=model,
    temperature=temperature,
    instructions=schema_obj.developer_system_prompt,
    input=doc_msg.openai_responses_input,
    text_format=schema_obj.inference_pydantic_model,
)


# Validate response and remove reasoning fields
extraction = schema_obj.pydantic_model.model_validate(filter_auxiliary_fields_json(response.output_text))

# Output
print("\n✅ Extracted Result (Responses API):")
print(extraction.model_dump_json(indent=2))

File: cookbook/node/documents/openai_responses_api.js
// ---------------------------------------------
// Full example: Use Retab with OpenAI's Responses API and JSON Schema extraction.
// ---------------------------------------------

import OpenAI from 'openai';
import { Schema, Retab } from '@retab/node';
import { filterAuxiliaryFieldsJson } from '@retab/node/utils/json_schema';
import { config } from 'dotenv';

// Load environment variables
config();

const apiKey = process.env.OPENAI_API_KEY;
const retabApiKey = process.env.RETAB_API_KEY;

if (!apiKey) {
  throw new Error('Missing OPENAI_API_KEY');
}
if (!retabApiKey) {
  throw new Error('Missing RETAB_API_KEY');
}

// Define schema
const jsonSchema = {
  'X-SystemPrompt': 'You are a useful assistant extracting information from documents.',
  properties: {
    name: {
      description: 'The name of the calendar event.',
      title: 'Name',
      type: 'string',
    },
    date: {
      'X-ReasoningPrompt': 'The user can mention it in any format, like **next week** or **tomorrow**. Infer the right date format from the user input.',
      description: 'The date of the calendar event in ISO 8601 format.',
      title: 'Date',
      type: 'string',
    },
  },
  required: ['name', 'date'],
  title: 'CalendarEvent',
  type: 'object',
};

// Configuration
const model = 'gpt-4o';
const modality = 'native';
const temperature = 0.0;
const imageResolutionDpi = 96;
const browserCanvas = 'A4';

// Retab Setup
const reclient = new Retab({ api_key: retabApiKey });
const docMsg = await reclient.documents.create_messages({
  document: '../../assets/docs/calendar_event.xlsx',
  // Note: Additional parameters like modality, image_resolution_dpi, browser_canvas 
  // would be supported by extending the create_messages implementation
});
const schemaObj = new Schema({ json_schema: jsonSchema });

// Transform the messages to OpenAI format
const openaiMessages = docMsg.openai_messages || [];

// OpenAI Responses API call
const client = new OpenAI({ apiKey });

// Note: The OpenAI Node.js SDK may not support the responses API in the same way as Python
// This is a conceptual example showing how it would work:

try {
  // Example 1: With hypothetical client.responses.create() method
  console.log('Note: OpenAI Responses API may not be available in Node.js SDK');
  console.log('Using regular completions API instead...');
  
  const response = await client.chat.completions.create({
    model,
    temperature,
    messages: [...schemaObj.openai_messages, ...openaiMessages],
    response_format: {
      type: 'json_schema',
      json_schema: {
        name: schemaObj.id,
        schema: schemaObj.inference_json_schema,
        strict: true,
      },
    },
  });

  // Validate response and remove reasoning fields
  if (!response.choices[0].message.content) {
    throw new Error('No content in response');
  }

  const parsedContent = JSON.parse(response.choices[0].message.content);
  const extraction = filterAuxiliaryFieldsJson(parsedContent);

  // Output
  console.log('\n✅ Extracted Result (Completions API):');
  console.log(JSON.stringify(extraction, null, 2));

} catch (error) {
  console.error('Error with API call:', error.message);
  console.log('Note: This example demonstrates the structure for OpenAI Responses API integration');
}